File: tray-ui/src/store.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const { app } = require('electron');

const MAX_EVENTS = 1000;

const DEFAULT_SETTINGS = {
  persistHistory: true,
  notificationsEnabled: true,
  notificationsSnoozedUntil: null, // null = off, 'indefinite' = until re-enabled, number = epoch ms
};

function userDataDir() {
  return app.getPath('userData');
}

function eventsPath() {
  return path.join(userDataDir(), 'events.json');
}

function settingsPath() {
  return path.join(userDataDir(), 'settings.json');
}

class Store {
  constructor() {
    this.settings = this._loadSettings();
    this.events = this.settings.persistHistory ? this._loadEvents() : [];
  }

  _loadSettings() {
    try {
      const raw = fs.readFileSync(settingsPath(), 'utf8');
      return { ...DEFAULT_SETTINGS, ...JSON.parse(raw) };
    } catch {
      return { ...DEFAULT_SETTINGS };
    }
  }

  _saveSettings() {
    try {
      fs.mkdirSync(userDataDir(), { recursive: true });
      fs.writeFileSync(settingsPath(), JSON.stringify(this.settings, null, 2));
    } catch {
    }
  }

  _loadEvents() {
    try {
      const raw = fs.readFileSync(eventsPath(), 'utf8');
      const arr = JSON.parse(raw);
      return Array.isArray(arr) ? arr : [];
    } catch {
      return [];
    }
  }

  _saveEvents() {
    if (!this.settings.persistHistory) return;
    try {
      fs.mkdirSync(userDataDir(), { recursive: true });
      fs.writeFileSync(eventsPath(), JSON.stringify(this.events.slice(0, MAX_EVENTS)));
    } catch {
    }
  }

  getSettings() {
    return { ...this.settings };
  }

  updateSettings(partial) {
    const wasPersisting = this.settings.persistHistory;
    this.settings = { ...this.settings, ...partial };
    this._saveSettings();

    if (!this.settings.persistHistory && wasPersisting) {
      try { fs.unlinkSync(eventsPath()); } catch { }
    }
    return this.getSettings();
  }

  addEvent(ev) {
    const withId = { ...ev, id: `${Date.now()}-${Math.random().toString(36).slice(2, 8)}` };
    this.events.unshift(withId);
    if (this.events.length > MAX_EVENTS) this.events.length = MAX_EVENTS;
    this._saveEvents();
    return withId;
  }

  clearEvents() {
    this.events = [];
    try { fs.unlinkSync(eventsPath()); } catch { }
  }

  getEvent(id) {
    return this.events.find((e) => e.id === id) ?? null;
  }

  getAllEvents() {
    return [...this.events];
  }

  notificationsSnoozed() {
    const until = this.settings.notificationsSnoozedUntil;
    if (!until) return false;
    if (until === 'indefinite') return true;
    return Date.now() < until;
  }

  queryEvents({ page = 1, pageSize = 20, search = '', correlatedOnly = false, sinceDays = null } = {}) {
    let filtered = this.events;

    if (search) {
      const q = search.toLowerCase();
      filtered = filtered.filter((e) =>
        (e.domain || '').toLowerCase().includes(q) ||
        (e.processName || '').toLowerCase().includes(q));
    }
    if (correlatedOnly) {
      filtered = filtered.filter((e) => e.correlatedFileAccess);
    }
    if (sinceDays) {
      const cutoff = Date.now() - sinceDays * 24 * 60 * 60 * 1000;
      filtered = filtered.filter((e) => new Date(e.timestamp).getTime() >= cutoff);
    }

    const total = filtered.length;
    const totalPages = Math.max(1, Math.ceil(total / pageSize));
    const clampedPage = Math.min(Math.max(1, page), totalPages);
    const start = (clampedPage - 1) * pageSize;

    return {
      items: filtered.slice(start, start + pageSize),
      total,
      page: clampedPage,
      pageSize,
      totalPages,
    };
  }
}

module.exports = { Store };

File: AntiGrabber.Tray/Store.cs
using System.Text.Json;
using AntiGrabber.Tray.Models;

namespace AntiGrabber.Tray;

public sealed record EventQuery(
    int Page = 1,
    int PageSize = 20,
    string? Search = null,
    bool CorrelatedOnly = false,
    int? SinceDays = null);

public sealed record EventQueryResult(
    IReadOnlyList<StoredBlockEvent> Items,
    int Total,
    int Page,
    int PageSize,
    int TotalPages);

public sealed record DailyBlockCount(string Date, int Count);
public sealed record TopCount(string Key, int Count);
public sealed record BlockStats(
    IReadOnlyList<DailyBlockCount> Daily,
    IReadOnlyList<TopCount> TopProcesses,
    IReadOnlyList<TopCount> TopDomains,
    int Total);

// Porta store.js pra C# 1:1: mesmo formato de settings.json/events.json, mesmo
// limite de 1000 eventos, engole erro de I/O em silêncio (mesmo comportamento
// do original — perder um write de settings não pode derrubar o app).
public sealed class Store
{
    private const int MaxEvents = 1000;
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    private readonly string _settingsPath;
    private readonly string _eventsPath;
    private readonly object _lock = new();

    private AppSettings _settings;
    private List<StoredBlockEvent> _events;

    public Store()
    {
        var dataDir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "AntiGrabber");
        Directory.CreateDirectory(dataDir);
        _settingsPath = Path.Combine(dataDir, "settings.json");
        _eventsPath = Path.Combine(dataDir, "events.json");

        _settings = LoadSettings();
        _events = _settings.PersistHistory ? LoadEvents() : new List<StoredBlockEvent>();
    }

    private AppSettings LoadSettings()
    {
        try
        {
            var json = File.ReadAllText(_settingsPath);
            return JsonSerializer.Deserialize<AppSettings>(json) ?? new AppSettings();
        }
        catch
        {
            return new AppSettings();
        }
    }

    private void SaveSettings()
    {
        try
        {
            File.WriteAllText(_settingsPath, JsonSerializer.Serialize(_settings, JsonOptions));
        }
        catch
        {
        }
    }

    private List<StoredBlockEvent> LoadEvents()
    {
        try
        {
            var json = File.ReadAllText(_eventsPath);
            return JsonSerializer.Deserialize<List<StoredBlockEvent>>(json) ?? new List<StoredBlockEvent>();
        }
        catch
        {
            return new List<StoredBlockEvent>();
        }
    }

    private void SaveEvents()
    {
        if (!_settings.PersistHistory) return;
        try
        {
            var toSave = _events.Count > MaxEvents ? _events.GetRange(0, MaxEvents) : _events;
            File.WriteAllText(_eventsPath, JsonSerializer.Serialize(toSave));
        }
        catch
        {
        }
    }

    public AppSettings GetSettings()
    {
        lock (_lock) return _settings;
    }

    public AppSettings UpdateSettings(Action<AppSettings> apply)
    {
        lock (_lock)
        {
            var wasPersisting = _settings.PersistHistory;
            apply(_settings);
            SaveSettings();

            if (!_settings.PersistHistory && wasPersisting)
            {
                try { File.Delete(_eventsPath); } catch { }
            }
            return _settings;
        }
    }

    public StoredBlockEvent AddEvent(StoredBlockEvent ev)
    {
        lock (_lock)
        {
            _events.Insert(0, ev);
            if (_events.Count > MaxEvents) _events.RemoveRange(MaxEvents, _events.Count - MaxEvents);
            SaveEvents();
            return ev;
        }
    }

    public void ClearEvents()
    {
        lock (_lock)
        {
            _events.Clear();
            try { File.Delete(_eventsPath); } catch { }
        }
    }

    public StoredBlockEvent? GetEvent(string id)
    {
        lock (_lock) return _events.FirstOrDefault(e => e.Id == id);
    }

    public IReadOnlyList<StoredBlockEvent> GetAllEvents()
    {
        lock (_lock) return _events.ToList();
    }

    public EventQueryResult QueryEvents(EventQuery query)
    {
        lock (_lock)
        {
            IEnumerable<StoredBlockEvent> filtered = _events;

            if (!string.IsNullOrWhiteSpace(query.Search))
            {
                var q = query.Search.Trim().ToLowerInvariant();
                filtered = filtered.Where(e =>
                    e.Domain.ToLowerInvariant().Contains(q) ||
                    e.ProcessName.ToLowerInvariant().Contains(q));
            }
            if (query.CorrelatedOnly)
            {
                filtered = filtered.Where(e => e.CorrelatedFileAccess);
            }
            if (query.SinceDays is { } days)
            {
                var cutoff = DateTimeOffset.UtcNow.AddDays(-days);
                filtered = filtered.Where(e => e.Timestamp >= cutoff);
            }

            var list = filtered.ToList();
            var total = list.Count;
            var pageSize = query.PageSize <= 0 ? 20 : query.PageSize;
            var totalPages = Math.Max(1, (int)Math.Ceiling(total / (double)pageSize));
            var page = Math.Min(Math.Max(1, query.Page), totalPages);
            var start = (page - 1) * pageSize;

            return new EventQueryResult(
                list.Skip(start).Take(pageSize).ToList(), total, page, pageSize, totalPages);
        }
    }

    public BlockStats GetStats(int days = 14)
    {
        lock (_lock)
        {
            var today = DateTimeOffset.UtcNow.Date;
            var daily = new List<DailyBlockCount>();
            for (var i = days - 1; i >= 0; i--)
            {
                var day = today.AddDays(-i);
                var count = _events.Count(e => e.Timestamp.UtcDateTime.Date == day);
                daily.Add(new DailyBlockCount(day.ToString("yyyy-MM-dd"), count));
            }

            var topProcesses = _events
                .GroupBy(e => e.ProcessName, StringComparer.OrdinalIgnoreCase)
                .Select(g => new TopCount(g.Key, g.Count()))
                .OrderByDescending(t => t.Count).Take(5).ToList();
            var topDomains = _events
                .GroupBy(e => e.Domain, StringComparer.OrdinalIgnoreCase)
                .Select(g => new TopCount(g.Key, g.Count()))
                .OrderByDescending(t => t.Count).Take(5).ToList();

            return new BlockStats(daily, topProcesses, topDomains, _events.Count);
        }
    }
}
